File: packages/silo-client/test/examples.test.ts
import { describe, expect, test } from "bun:test";
import { Filter } from "../src/query/filter";
import { Sort } from "../src/query/sort";
import { EnvironmentHandle } from "../src/scope/environment-handle";
import { ProjectHandle } from "../src/scope/project-handle";
import { FetchTransport as Transport } from "../src/transport/fetch-transport";
import { StubFetch } from "./support/stub-fetch";
import { StubResponse } from "./support/stub-response";

/**
 * The README's examples, each block
 * builds the same handle chain `Silo.project().environment()` would, and
 * proves it compiles and runs exactly as documented.
 */
interface Movie {
  title: string;
  year: number;
  status: "draft" | "published";
  genres: string[];
  trailerUrl?: string;
}

describe("Projects and environments", () => {
  test("rename is bound to the identity reviewed in a dry run", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(
      StubResponse.json({ id: "01J8", from: "moviespace", to: "movie-space", rewritten_claims: [], pattern_affected_claims: [] }),
    );
    stubFetch.enqueue(
      StubResponse.json({ id: "01J8", from: "moviespace", to: "movie-space", rewritten_claims: [], pattern_affected_claims: [] }),
    );
    stubFetch.enqueue(StubResponse.empty());

    const moviespace = new ProjectHandle(new Transport({ url: "http://localhost:8090", fetch: stubFetch.fetch }), "moviespace");

    const preview = await moviespace.rename("movie-space", { dryRun: true });
    preview.id;
    preview.rewrittenClaims;
    preview.patternAffectedClaims;
    await moviespace.rename("movie-space", { expectedId: preview.id });
    await moviespace.delete({ force: true });

    expect(stubFetch.received).toHaveLength(3);
  });
});

describe("Collections and schemas", () => {
  test("create, schema get/put, rename", async () => {
    const stubFetch = new StubFetch();
    const schema = {
      type: "object",
      required: ["title"],
      properties: { title: { type: "string" }, year: { type: "integer" }, status: { type: "string" } },
    };
    stubFetch.enqueue(StubResponse.json({ id: "01J8", name: "movies", schema }, 201));
    stubFetch.enqueue(StubResponse.json({ id: "01J8", name: "movies", schema }));
    stubFetch.enqueue(StubResponse.json({ id: "01J8", name: "movies", schema }));

    const transport = new Transport({ url: "http://localhost:8090", fetch: stubFetch.fetch });
    const environment = new ProjectHandle(transport, "moviespace").environment("prod");

    await environment.collections.create("movies", schema);
    const movies = environment.collection<Movie>("movies");
    await movies.schema.get();
    await movies.schema.put(schema);

    expect(stubFetch.received).toHaveLength(3);
  });
});

describe("Entries: reading, editing, writing, listing", () => {
  const environmentOf = (stubFetch: StubFetch): EnvironmentHandle => {
    const transport = new Transport({ url: "http://localhost:8090", fetch: stubFetch.fetch });
    return new ProjectHandle(transport, "moviespace").environment("prod");
  };

  test("a read answers a flat row, and writes take its id and rev", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(
      StubResponse.json({
        id: "01J8",
        rev: 3,
        created_at: "2026-01-01T00:00:00.000Z",
        updated_at: "2026-01-01T00:00:00.000Z",
        title: "Arrival",
        year: 2016,
        status: "draft",
        genres: ["sci-fi"],
        trailerUrl: "https://cdn.moviespace.com/trailers/arrival.mp4",
      }),
    );
    stubFetch.enqueue(StubResponse.empty());
    const movies = environmentOf(stubFetch).collection<Movie>("movies");

    const movie = await movies.get("01J8");

    expect(movie.trailerUrl).toBe("https://cdn.moviespace.com/trailers/arrival.mp4");
    // @ts-expect-error -- a row carries no methods, proven through the full
    // ProjectHandle -> EnvironmentHandle -> CollectionHandle chain.
    movie.save;
    await movies.delete(movie.id, movie.rev);
  });

  test("a raw read carries the template, and replace() writes it back untouched", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(
      StubResponse.json({
        id: "01J8",
        rev: 3,
        created_at: "2026-01-01T00:00:00.000Z",
        updated_at: "2026-01-01T00:00:00.000Z",
        title: "Arrival",
        year: 2016,
        status: "draft",
        genres: ["sci-fi"],
        trailerUrl: "{{CDN_URL}}/trailers/arrival.mp4",
      }),
    );
    stubFetch.enqueue(
      StubResponse.json({
        id: "01J8",
        rev: 4,
        created_at: "2026-01-01T00:00:00.000Z",
        updated_at: "2026-01-02T00:00:00.000Z",
        title: "Arrival",
        year: 2016,
        status: "published",
        genres: ["sci-fi"],
        trailerUrl: "{{CDN_URL}}/trailers/arrival.mp4",
      }),
    );
    const movies = environmentOf(stubFetch).collection<Movie>("movies");

    const draft = await movies.get("01J8", { variables: "raw" });
    expect(draft.trailerUrl).toBe("{{CDN_URL}}/trailers/arrival.mp4");

    const { id, rev, created_at, updated_at, ...fields } = draft;
    await movies.replace(id, rev, { ...fields, status: "published" });

    // The template survives the round trip, which is the whole point of
    // reading raw before writing.
    expect(JSON.parse(stubFetch.received[1].body ?? "{}").trailerUrl).toBe("{{CDN_URL}}/trailers/arrival.mp4");
    void [created_at, updated_at];
  });

  test("a row is a plain value: it clones, and carries nothing alongside it", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(
      StubResponse.json({
        id: "01J8",
        rev: 1,
        created_at: "2026-01-01T00:00:00.000Z",
        updated_at: "2026-01-01T00:00:00.000Z",
        title: "Arrival",
        year: 2016,
        status: "draft",
        genres: ["sci-fi"],
      }),
    );
    const movies = environmentOf(stubFetch).collection<Movie>("movies");

    const movie = await movies.get("01J8");

    // No transport, no scope, no prototype: what a store or a structured
    // clone would have choked on before (D62).
    expect(Object.keys(movie).sort()).toEqual(
      ["created_at", "genres", "id", "rev", "status", "title", "updated_at", "year"],
    );
    expect(structuredClone(movie)).toEqual(movie);
    expect(JSON.parse(JSON.stringify(movie))).toEqual(movie);
  });

  test("create sends ?variables=raw and answers the row it stored", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(
      StubResponse.json(
        { id: "01J9", rev: 1, created_at: "2026-01-01T00:00:00.000Z", updated_at: "2026-01-01T00:00:00.000Z", title: "Arrival", year: 2016, status: "draft", genres: ["sci-fi"] },
        201,
      ),
    );
    stubFetch.enqueue(StubResponse.empty());
    const movies = environmentOf(stubFetch).collection<Movie>("movies");

    const created = await movies.create({ title: "Arrival", year: 2016, status: "draft", genres: ["sci-fi"] });
    expect(created.rev).toBe(1);
    await movies.delete(created.id, created.rev);
  });

  test("list() answers a page with the server's window, and next() advances by it", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(
      StubResponse.json({
        data: [
          { id: "01J8", rev: 1, created_at: "2026-01-01T00:00:00.000Z", updated_at: "2026-01-01T00:00:00.000Z", title: "Arrival", year: 2016, status: "published", genres: ["sci-fi"] },
        ],
        total: 137,
        limit: 500,
        offset: 0,
      }),
    );
    const movies = environmentOf(stubFetch).collection<Movie>("movies");

    const filtered = await movies.list({
      where: movies.filter.field("status").equals("published").and(movies.filter.each("genres").equals("sci-fi")).and(movies.filter.field("title").contains("arrival")),
      sort: Sort.recentlyUpdated(),
      limit: 900,
    });

    expect(filtered.limit).toBe(500);
    expect(filtered.total).toBe(137);
    expect(filtered.pageNumber).toBe(1);
    expect(filtered.hasMore).toBe(true);
  });
});

describe("Queries: filters and sort, untyped statics", () => {
  test("every leaf operator builds the documented wire shape", () => {
    expect(Filter.field("status").equals("published").toJSON()).toEqual({ op: "eq", path: "$.data.status", value: "published" });
    expect(Filter.field("status").notEquals("draft").toJSON()).toEqual({ op: "neq", path: "$.data.status", value: "draft" });
    expect(Filter.field("title").contains("arrival").toJSON()).toEqual({ op: "contains", path: "$.data.title", value: "arrival" });
    expect(Filter.field("year").greaterThan(2000).toJSON()).toEqual({ op: "gt", path: "$.data.year", value: 2000 });
    expect(Filter.field("year").atLeast(2000).toJSON()).toEqual({ op: "gte", path: "$.data.year", value: 2000 });
    expect(Filter.field("year").lessThan(2000).toJSON()).toEqual({ op: "lt", path: "$.data.year", value: 2000 });
    expect(Filter.field("year").atMost(2000).toJSON()).toEqual({ op: "lte", path: "$.data.year", value: 2000 });
    expect(Filter.field("status").oneOf(["draft", "review"]).toJSON()).toEqual({ op: "in", path: "$.data.status", value: ["draft", "review"] });
    expect(Filter.field("tagline").exists().toJSON()).toEqual({ op: "exists", path: "$.data.tagline" });
    expect(Filter.each("genres").notEquals("horror").toJSON()).toEqual({ op: "neq", path: "$.data.genres[*]", value: "horror" });
    expect(Filter.not(Filter.each("genres").equals("horror")).toJSON()).toEqual({
      op: "not",
      args: [{ op: "eq", path: "$.data.genres[*]", value: "horror" }],
    });
    // A dot reaches inside a nested field, as the README's `director.name`.
    expect(Filter.field("director.name").contains("villeneuve").toJSON()).toEqual({
      op: "contains",
      path: "$.data.director.name",
      value: "villeneuve",
    });
  });

  test("Sort statics", () => {
    expect(Sort.by("title").toString()).toBe("$.data.title");
    expect(Sort.by("title").descending().toString()).toBe("-$.data.title");
    expect(Sort.meta("created_at").ascending().toString()).toBe("$.created_at");
    expect(Sort.recentlyUpdated().toString()).toBe("-$.updated_at");
    expect(Sort.recentlyCreated().toString()).toBe("-$.created_at");
    expect(Sort.of(Sort.by("status"), Sort.recentlyUpdated())).toBe("$.data.status,-$.updated_at");
  });
});

describe("Variables", () => {
  test("declare, rename, describe, undeclare, list, set, unset", async () => {
    const stubFetch = new StubFetch();
    const declaration = (overrides: Record<string, unknown> = {}) => ({
      name: "CDN_URL",
      description: "Public asset root",
      value: "https://cdn.moviespace.com",
      set_in: 1,
      created_at: "2026-01-01T00:00:00.000Z",
      updated_at: "2026-01-01T00:00:00.000Z",
      ...overrides,
    });
    stubFetch.enqueue(StubResponse.json(declaration(), 201));
    stubFetch.enqueue(StubResponse.json(declaration({ name: "PUBLIC_CDN_URL" })));
    stubFetch.enqueue(StubResponse.json(declaration({ name: "PUBLIC_CDN_URL", description: "The public asset root" })));
    stubFetch.enqueue(StubResponse.empty());
    stubFetch.enqueue(StubResponse.json({ items: [declaration()] }));
    stubFetch.enqueue(StubResponse.json(declaration()));
    stubFetch.enqueue(StubResponse.json(declaration({ value: null, set_in: 0 })));

    const transport = new Transport({ url: "http://localhost:8090", fetch: stubFetch.fetch });
    const moviespace = new ProjectHandle(transport, "moviespace");

    await moviespace.variables.declare("CDN_URL", { description: "Public asset root", environment: "prod", value: "https://cdn.moviespace.com" });
    await moviespace.variables.rename("CDN_URL", "PUBLIC_CDN_URL");
    await moviespace.variables.describe("PUBLIC_CDN_URL", "The public asset root");
    await moviespace.variables.undeclare("PUBLIC_CDN_URL");

    const environment = moviespace.environment("prod");
    await environment.variables.list();
    await environment.variables.set("PUBLIC_CDN_URL", "https://cdn.moviespace.com");
    await environment.variables.unset("PUBLIC_CDN_URL");

    expect(stubFetch.received).toHaveLength(7);
  });
});

describe("Search", () => {
  test("one collection, one environment", async () => {
    const stubFetch = new StubFetch();
    stubFetch.enqueue(StubResponse.json({ data: [], total: 0, limit: 50, offset: 0, truncated: false, engine: "scan" }));
    stubFetch.enqueue(StubResponse.json({ data: [], total: 0, limit: 50, offset: 0, truncated: false, engine: "scan" }));

    const transport = new Transport({ url: "http://localhost:8090", fetch: stubFetch.fetch });
    const environment = new ProjectHandle(transport, "moviespace").environment("prod");
    const movies = environment.collection<Movie>("movies");

    await movies.search({ query: "arrival" });
    await environment.search({ query: "arrival" });

    expect(stubFetch.received).toHaveLength(2);
  });
});

File: packages/silo-client-java/src/test/java/in/org/quicko/silo/client/ExamplesTest.java
package in.org.quicko.silo.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import in.org.quicko.silo.client.collections.CollectionDefinition;
import in.org.quicko.silo.client.collections.CollectionHandle;
import in.org.quicko.silo.client.collections.CollectionSummary;
import in.org.quicko.silo.client.collections.JsonSchema;
import in.org.quicko.silo.client.entries.Entry;
import in.org.quicko.silo.client.scope.DeleteOptions;
import in.org.quicko.silo.client.scope.Environment;
import in.org.quicko.silo.client.scope.EnvironmentHandle;
import in.org.quicko.silo.client.scope.Project;
import in.org.quicko.silo.client.scope.ProjectHandle;
import in.org.quicko.silo.client.scope.RenameOptions;
import in.org.quicko.silo.client.scope.RenameReport;
import in.org.quicko.silo.client.search.SearchEngine;
import in.org.quicko.silo.client.search.SearchHit;
import in.org.quicko.silo.client.search.SearchPage;
import in.org.quicko.silo.client.search.SearchQuery;
import in.org.quicko.silo.client.support.StubHttp;
import in.org.quicko.silo.client.variables.DeclareVariableOptions;
import in.org.quicko.silo.client.variables.Variable;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * The README's examples, run.
 *
 * <p>Each test is one block of `README.md`, typed out as the README types it,
 * so an example cannot drift from the API without the build saying so. The
 * media blocks are in {@link MediaExamplesTest} for the reason the README puts
 * them in a section of their own.
 */
class ExamplesTest {

  /** The README's field type, a record, which is also the claim that one works. */
  record Post(String title, String status, List<String> tags) {}

  private static final String Schema = """
      {"type":"object","required":["title","status"],"properties":{
        "title":{"type":"string"},
        "status":{"type":"string","enum":["draft","published"]},
        "tags":{"type":"array","items":{"type":"string"}}}}
      """;

  private StubHttp server;

  @BeforeEach
  void newStub() {
    server = StubHttp.create();
  }

  private static String row(String id, int rev, String status) {
    return "{\"id\":\"" + id + "\",\"rev\":" + rev
        + ",\"created_at\":\"2026-09-01T10:00:00.000Z\""
        + ",\"updated_at\":\"2026-09-02T10:00:00.000Z\""
        + ",\"title\":\"Hello\",\"status\":\"" + status + "\",\"tags\":[\"intro\"]}";
  }

  @Test
  void gettingStarted() {
    server.enqueueJson("{\"status\":\"ok\",\"version\":\"1.1.0\"}");
    server.enqueueJson("{\"id\":\"01C\",\"name\":\"posts\",\"schema\":" + Schema + "}");
    server.enqueueJson(row("01ABC", 1, "draft"));
    server.enqueueJson(row("01ABC", 2, "published"));
    server.enqueueJson("{\"data\":[" + row("01ABC", 2, "published") + "],\"total\":1,\"limit\":50,\"offset\":0}");
    server.enqueueNoContent();

    Silo silo = server.silo("silo_key_abc");

    assertEquals("1.1.0", silo.health().version());

    EnvironmentHandle prod = silo.scope("acme", "prod");
    prod.collections().create("posts", JsonSchema.parse(Schema));

    CollectionHandle<Post> posts = prod.collection("posts", Post.class);

    Entry<Post> created = posts.create(new Post("Hello", "draft", List.of("intro")));
    Entry<Post> published = posts.replace(
        created.id(), created.rev(), new Post("Hello", "published", List.of("intro")));

    List<String> titles = new ArrayList<>();
    for (Entry<Post> post : posts.all()) {
      titles.add(post.fields().title());
    }

    posts.delete(published.id(), published.rev());

    assertEquals("draft", created.fields().status(), "a record binds as the field type");
    assertEquals(2, published.rev());
    assertEquals(List.of("Hello"), titles);
    assertEquals(6, server.requestCount());
  }

  @Test
  void projectsAndEnvironments() {
    server.enqueueJson("{\"items\":[{\"id\":\"01P\",\"name\":\"acme\"}]}");
    server.enqueueJson("{\"id\":\"01P\",\"name\":\"acme\"}");
    server.enqueueJson("{\"items\":[{\"id\":\"01E\",\"name\":\"prod\"}]}");
    server.enqueueJson("{\"id\":\"01E2\",\"name\":\"staging\"}");

    Silo silo = server.silo("k");

    List<Project> projects = silo.projects().list();
    Project acme = silo.projects().create("acme");

    ProjectHandle project = silo.project("acme");
    List<Environment> environments = project.environments().list();
    project.environments().create("staging");

    assertEquals("acme", projects.get(0).name());
    assertEquals("01P", acme.id());
    assertEquals("prod", environments.get(0).name());
  }

  @Test
  void aRenameIsPreviewedThenBoundToWhatWasPreviewed() {
    String report = "{\"id\":\"01P\",\"from\":\"acme\",\"to\":\"acme-corp\","
        + "\"rewritten_claims\":[\"collections:acme/*/*:entries:read\"],"
        + "\"pattern_affected_claims\":[\"collections:acm*/*/*:entries:read\"]}";
    server.enqueueJson(report);
    server.enqueueJson(report);
    server.enqueueNoContent();

    ProjectHandle project = server.silo("k").project("acme");

    RenameReport preview = project.rename("acme-corp", RenameOptions.preview());
    assertEquals(List.of("collections:acme/*/*:entries:read"), preview.rewrittenClaims());
    assertEquals(List.of("collections:acm*/*/*:entries:read"), preview.patternAffectedClaims());

    project.rename("acme-corp", RenameOptions.none().expectedId(preview.id()));

    assertTrue(server.request(0).target().contains("dry_run=true"));
    assertTrue(server.request(1).target().contains("expected_id=01P"));
    assertTrue(!server.request(1).target().contains("dry_run"));

    server.silo("k").project("acme").environment("staging").delete(DeleteOptions.forced());
    assertEquals("/api/projects/acme/envs/staging?force=true", server.lastRequest().target());
  }

  @Test
  void collectionsAndSchemas() {
    server.enqueueJson("{\"items\":[{\"id\":\"01C\",\"name\":\"posts\",\"entries\":42,"
        + "\"requires_auth\":true,\"created_at\":\"2026-09-01T10:00:00.000Z\","
        + "\"updated_at\":\"2026-09-02T10:00:00.000Z\"}]}");
    server.enqueueJson("{\"id\":\"01C\",\"name\":\"posts\",\"schema\":" + Schema + "}");
    server.enqueueJson("{\"items\":[{\"id\":\"01C\",\"name\":\"posts\",\"schema\":" + Schema + "}]}");
    server.enqueueJson("{\"id\":\"01C\",\"name\":\"posts\",\"schema\":" + Schema + "}");
    server.enqueueNoContent();

    EnvironmentHandle prod = server.silo("k").scope("acme", "prod");

    List<CollectionSummary> summaries = prod.collections().list();
    assertEquals(42, summaries.get(0).entries());

    CollectionDefinition definition = prod.collection("posts").schema().get();
    List<CollectionDefinition> everySchema = prod.schemas();
    assertEquals("posts", definition.name());
    assertEquals(1, everySchema.size());

    prod.collection("posts").schema().put(JsonSchema.parse(Schema));
    prod.collection("posts").schema().delete(DeleteOptions.forced());

    assertEquals("/api/projects/acme/envs/prod/collections/posts/schema?force=true",
        server.lastRequest().target());
  }

  @Test
  void variablesAreDeclaredOncePerProjectAndValuedPerEnvironment() {
    String declaration = "{\"name\":\"API_URL\",\"description\":\"Where the public API lives\","
        + "\"value\":\"https://api.acme.com\",\"set_in\":1,"
        + "\"created_at\":\"2026-09-01T10:00:00.000Z\",\"updated_at\":\"2026-09-01T10:00:00.000Z\"}";
    server.enqueueJson(declaration);
    server.enqueueJson(declaration.replace("https://api.acme.com", "https://api.staging.acme.com"));
    server.enqueueJson("{\"items\":[" + declaration + ",{\"name\":\"BANNER\",\"description\":\"\","
        + "\"value\":null,\"set_in\":0,\"created_at\":\"2026-09-01T10:00:00.000Z\","
        + "\"updated_at\":\"2026-09-01T10:00:00.000Z\"}]}");
    server.enqueueJson(declaration);
    server.enqueueNoContent();

    Silo silo = server.silo("k");

    silo.project("acme").variables().declare("API_URL",
        DeclareVariableOptions.none()
            .description("Where the public API lives")
            .environment("prod")
            .value("https://api.acme.com"));

    silo.scope("acme", "staging").variables().set("API_URL", "https://api.staging.acme.com");

    List<Variable> variables = silo.scope("acme", "prod").variables().list();
    assertEquals("https://api.acme.com", variables.get(0).value().orElse("(unset)"));
    assertEquals("(unset)", variables.get(1).value().orElse("(unset)"),
        "an unset value leaves the reference standing, and is not an empty string");

    silo.scope("acme", "staging").variables().unset("API_URL");
    silo.project("acme").variables().undeclare("API_URL");

    assertEquals("/api/projects/acme/variables/API_URL", server.lastRequest().target());
  }

  @Test
  void searchReachIsTheReceiver() {
    String hit = "{\"project\":\"acme\",\"env\":\"prod\",\"collection\":\"posts\",\"entry\":"
        + row("01ABC", 1, "published") + ",\"snippets\":[{\"path\":\"$.data.title\","
        + "\"before\":\"\",\"match\":\"release notes\",\"after\":\" for 1.1\"}]}";
    String body = "{\"data\":[" + hit + "],\"total\":1,\"limit\":20,\"offset\":0,\"engine\":\"fts5\"}";
    server.enqueueJson(body);
    server.enqueueJson(body);
    server.enqueueJson(body);

    Silo silo = server.silo("k");
    SearchQuery query = SearchQuery.matching("release notes").limit(20);

    SearchPage page = silo.search(query);
    for (SearchHit found : page.hits()) {
      assertEquals("prod", found.environment());
      assertEquals("release notes", found.snippets().get(0).match());
    }

    silo.scope("acme", "prod").search(query);
    silo.scope("acme", "prod").collection("posts").search(query);

    assertEquals("/api/search", server.request(0).path());
    assertEquals("/api/projects/acme/envs/prod/search", server.request(1).path());
    assertEquals("/api/projects/acme/envs/prod/collections/posts/search", server.request(2).path());
  }

  @Test
  void aTruncatedSearchWillNotGuessAPageCount() {
    server.enqueueJson("{\"data\":[],\"total\":0,\"limit\":20,\"offset\":0,"
        + "\"engine\":\"scan\",\"truncated\":true}");

    SearchPage page = server.silo("k").search(SearchQuery.matching("release notes").limit(20));

    assertTrue(page.truncated());
    assertTrue(page.pageCount().isEmpty(), "the total is a scan's count, not a real one");
    assertEquals(SearchEngine.SCAN, page.engine());
  }
}
